src/docfx/lib/msgraph/MicrosoftGraphCache.cs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class MicrosoftGraphCache : IDisposable
+    {
+        private readonly string _cachePath;
+        private readonly double _expirationInHours;
+        private readonly SemaphoreSlim _syncRoot = new SemaphoreSlim(1, 1);
+        private readonly MicrosoftGraphAccessor _microsoftGraphAccessor = null;
+        private Dictionary<string, MicrosoftAlias> _aliases = new Dictionary<string, MicrosoftAlias>();
+        private IList<string> _tempInvalidAliases = new List<string>();
+        private bool _needUpdate = false;
+
+        public MicrosoftGraphCache(Config config)
+        {
+            _cachePath = AppData.MicrosoftGraphCachePath;
+            _expirationInHours = config.MicrosoftGraph.MicrosoftAliasCacheExpirationInHours;
+
+            if (!string.IsNullOrEmpty(config.MicrosoftGraph.MicrosoftGraphTenantId) &&
+                !string.IsNullOrEmpty(config.MicrosoftGraph.MicrosoftGraphClientId) &&
+                !string.IsNullOrEmpty(config.MicrosoftGraph.MicrosoftGraphClientSecret))
+            {
+                _microsoftGraphAccessor = new MicrosoftGraphAccessor(
+                    config.MicrosoftGraph.MicrosoftGraphTenantId,
+                    config.MicrosoftGraph.MicrosoftGraphClientId,
+                    config.MicrosoftGraph.MicrosoftGraphClientSecret);
+            }
+
+            if (File.Exists(_cachePath))
+            {
+                var cacheFile = JsonUtility.Deserialize<MicrosoftGraphCacheFile>(ProcessUtility.ReadFile(_cachePath), _cachePath);
+                _aliases = cacheFile.Aliases.Where(msAlias => msAlias.Expiry >= DateTime.UtcNow).ToDictionary(msAlias => msAlias.Alias);
+            }
+        }
+
+        public Task<(Error error, MicrosoftAlias msAlias)> GetMicrosoftAliasAsync(string alias)
+        {
+            if (string.IsNullOrEmpty(alias))
+            {
+                return new Task<(Error error, MicrosoftAlias msAlias)>(null, null);
+            }
+
+            return Synchronized(GetAsyncCore);
+
+            async Task<(Error error, MicrosoftAlias msAlias)> GetAsyncCore()
+            {
+                if (_aliases.TryGetValue(alias, out var msAlias))
+                {
+                    return (null, msAlias);
+                }
+
+                if (_microsoftGraphAccessor != null)
```

Comment:
nit: consider early return to reduce indention

```suggestion
                if (_microsoftGraphAccessor == null)
                {
                    return default;
                }
```